File: src/main/java/com/viaversion/aas/codec/CryptoCodec.java
package com.viaversion.aas.codec;

import io.netty.buffer.ByteBuf;
import io.netty.channel.ChannelHandlerContext;
import io.netty.handler.codec.MessageToMessageCodec;
import org.jetbrains.annotations.NotNull;

import javax.crypto.Cipher;
import java.util.List;

public class CryptoCodec extends MessageToMessageCodec<ByteBuf, ByteBuf> {
	private final Cipher cipherEncode;
	private final Cipher cipherDecode;

	public CryptoCodec(@NotNull Cipher cipherEncode, @NotNull Cipher cipherDecode) {
		this.cipherEncode = cipherEncode;
		this.cipherDecode = cipherDecode;
	}

	@Override
	protected void encode(ChannelHandlerContext ctx, ByteBuf msg, List<Object> out) throws Exception {
		var i = msg.readerIndex();
		var size = msg.readableBytes();
		msg.writerIndex(i + cipherEncode.update(msg.nioBuffer(), msg.nioBuffer(i, cipherEncode.getOutputSize(size))));
		out.add(msg.retain());
	}

	@Override
	protected void decode(ChannelHandlerContext ctx, ByteBuf msg, List<Object> out) throws Exception {
		if (!ctx.channel().isActive()) return;
		var i = msg.readerIndex();
		var size = msg.readableBytes();
		msg.writerIndex(i + cipherDecode.update(msg.nioBuffer(), msg.nioBuffer(i, cipherDecode.getOutputSize(size))));
		out.add(msg.retain());
	}
}

File: src/main/kotlin/com/viaversion/aas/codec/CryptoCodec.kt
package com.viaversion.aas.codec

import io.netty.buffer.ByteBuf
import io.netty.channel.ChannelHandlerContext
import io.netty.handler.codec.MessageToMessageCodec
import javax.crypto.Cipher

class CryptoCodec(val cipherDecode: Cipher, var cipherEncode: Cipher) : MessageToMessageCodec<ByteBuf, ByteBuf>() {
    override fun decode(ctx: ChannelHandlerContext, msg: ByteBuf, out: MutableList<Any>) {
        if (!ctx.channel().isActive) return
        val i = msg.readerIndex()
        val size = msg.readableBytes()
        msg.writerIndex(i + cipherDecode.update(msg.nioBuffer(), msg.nioBuffer(i, cipherDecode.getOutputSize(size))))
        out.add(msg.retain())
    }

    override fun encode(ctx: ChannelHandlerContext, msg: ByteBuf, out: MutableList<Any>) {
        val i = msg.readerIndex()
        val size = msg.readableBytes()
        msg.writerIndex(i + cipherEncode.update(msg.nioBuffer(), msg.nioBuffer(i, cipherEncode.getOutputSize(size))))
        out.add(msg.retain())
    }
}